docs for quarkus (#984)
localstack/docs: content/en/user-guide/integrations/quarkus/index.md

---
title: "Quarkus"
weight: 11
description: >
  Use the Quarkus framework with LocalStack
---

Quarkus is a Java framework optimized for cloud, serverless, and containerized environments. Quarkus leverages a Kubernetes Native Java stack tailored for GraalVM & OpenJDK HotSpot, which further builds on various Java libraries and standards.

Localstack is supported by Quarkus as a Dev service for Amazon Services. Quarkus Amazon Services automatically starts a LocalStack container in development mode and when running tests, and the extension client is configured automatically.

> {{< alert title="Note" >}}

Reviewer: nit: seems like quite a specific detail - this is the first mention of `pom.xml` in the text and at this point the user doesn't even know about the Maven setup yet. Perhaps it could be moved a bit further down the page?

Author: Sure, will do 👍

- [Docker](https://docs.docker.com/get-docker/)

### Create a Maven project

Create a new project with the following command:

{{< command >}}
$ mvn io.quarkus.platform:quarkus-maven-plugin:3.6.3:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=amazon-lambda-quickstart \
    -DclassName="org.acme.lambda.QuarkusLambdaSyncResource" \
    -Dpath="/sync" \
    -Dextensions="resteasy-reactive-jackson,amazon-lambda"
$ cd amazon-lambda-quickstart
{{< /command >}}

The above command generates a Maven project structure with imports for RESTEasy Reactive/JAX-RS and Amazon Lambda Client extensions.

### Configure Lambda Client

Both Lambda clients (sync and async) can be configured through the `application.properties` file, which should be located in the `src/main/resources` directory. Additionally, ensure that a suitable implementation of the sync client is added to the `classpath`. By default, the extension employs the URL connection HTTP client, so it's necessary to include a URL connection client dependency in the `pom.xml` file:

```xml
<dependency>
    <groupId>software.amazon.awssdk</groupId>
    <artifactId>url-connection-client</artifactId>
</dependency>
```

> If you want to use Apache HTTP client instead, configure it as follows:

Reviewer: ```suggestion
If you want to use Apache HTTP client instead, configure it as follows in `application.properties`:
```

{{< command >}}
$ java -Dparameters.path=/quarkus/is/awesome/ -jar target/quarkus-app/quarkus-run.jar
{{< /command >}}

{{< alert title="Note" >}}
With GraalVM installed, you can also create a native executable binary using the following command:
{{< command >}}
$ ./mvnw clean package -Dnative.
{{< /command >}}
{{< /alert >}}

## Supported extensions

- [Lambda](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-lambda.html)
- [S3](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-s3.html)
- [SSM](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-ssm.html)
- [SQS](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-sqs.html)
- [SNS](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-sns.html)
- [SES](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-ses.html)
- [Secrets Manager](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-secretsmanager.html)
- [KMS](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-kms.html)
- [IAM](https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-iam.html)

## Configuration

> The Configuration properties are fixed at build time. All the other configuration properties are overridable at runtime.

Reviewer: ```suggestion
The following configuration properties are fixed at build time. All the other configuration properties can be overridden at runtime.
```

| `quarkus.aws.devservices.localstack.init-scripts-folder`  | string                 |                                      |
| `quarkus.aws.devservices.localstack.init-scripts-classpath`| string                 |                                      |
| `quarkus.aws.devservices.localstack.init-completion-msg`  | string                 |                                      |
| `quarkus.aws.devservices.localstack.container-properties` | `Map<String,String>`  |                                      |
| `quarkus.aws.devservices.localstack.additional-services."additional-services".enabled` | boolean                |                                      |
| `quarkus.aws.devservices.localstack.additional-services."additional-services".shared`  | boolean                | `false`                              |
| `quarkus.aws.devservices.localstack.additional-services."additional-services".service-name` | string                 | `localstack`                         |
| `quarkus.aws.devservices.localstack.additional-services."additional-services".container-properties` | `Map<String,String>`  |                                      |

{{< alert title="Note" >}}

> - If `quarkus.aws.devservices.localstack.additional-services."additional-services".enabled` is set to `true` and the endpoint-override is not configured, LocalStack will be started and utilized instead of the provided configuration. For all services excluding Cognito, LocalStack will function as the core cloud emulator.  In the case of Cognito, the emulation/mocking will be done by Moto.

Reviewer: Not sure whether the additional information related to Moto is required - seems like an implementation detail and not really necessary information for the user, wdyt? 

Also, curious why we're using Cognito from moto - presumably to allow users to run local tests even with the community version..? If that is the case, then we could maybe mention that Moto provides a basic version, whereas the full-fledged features are provided when using LS Pro?

Author: > Also, curious why we're using Cognito from moto

We aren't recommending it 😅 Quarkus uses Moto for Cognito for some reason, and not LocalStack. It maybe because Cognito falls under LocalStack Pro.

- The `quarkus.aws.devservices.localstack.additional-services."additional-services".shared` indicates whether the LocalStack container managed by Dev Services is shared. In shared mode, Quarkus utilizes label-based service discovery, specifically the `quarkus-dev-service-localstack` label, to identify running containers. If a matching container is found, it is used. Otherwise, Dev Services initiates a new container. It's important to note that sharing is not supported for the Cognito extension.

> - In `quarkus.aws.devservices.localstack.additional-services."additional-services".service-name`, the value of the `quarkus-dev-service-localstack` label is attached to the initiated container. In dev mode, when the shared flag is true, Dev Services checks for a container with the quarkus-dev-service-localstack label set to the configured value before starting a new one. If found, it utilizes the existing container. Otherwise, it creates a new container with the `quarkus-dev-service-localstack` label set to the specified value. In test mode, Dev Services groups services with the same service-name value into a single container instance. This property is useful when there's a requirement for multiple shared LocalStack instances.

Reviewer: ```suggestion
- In `quarkus.aws.devservices.localstack.additional-services."additional-services".service-name`, the value of the `quarkus-dev-service-localstack` label is attached to the initiated container. In dev mode, when the shared flag is true, Dev Services checks for a container with the `quarkus-dev-service-localstack` label set to the configured value before starting a new one. If found, it utilizes the existing container. Otherwise, it creates a new container with the `quarkus-dev-service-localstack` label set to the specified value. In test mode, Dev Services groups services with the same service-name value into a single container instance. This property is useful when there's a requirement for multiple shared LocalStack instances.
```
